### ISSUE TITLE
token invalidation timestamp can be misleading

### ISSUE THREAD

**Issue by hzshlomi** [relayrides/pushy#674]
Hi,

According to the documentation in:
https://github.com/relayrides/pushy/blob/master/pushy/src/main/java/com/turo/pushy/apns/PushNotificationResponse.java
The getTokenInvalidationTimestamp() method returns a Date object.

Yet, there is a contradiction between the method description and the return value description regarding what this Date really represents.

The method description says:
1. "returns the **most recent** time at which the APNs gateway **confirmed** that the token is no longer valid"

While the return value description says:
2. "the time at which the token for the sent push notification **became invalid**"

Now comes the APNS doc from here:
https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/sending_notification_requests_to_apns#2947616
Where (See table 4) "timestamp" is described as "The time at which APNs confirmed the token was no longer valid for the topic" which does not use the "most recent" wording, which basically gives a 3rd option:

3. "Any time between the last time the token was invalidated, and the last time the invalidation was confirmed"

The importance of the distinction comes from the action that follows this response: you are expected by APNS to remove that token from your DB, and depending on the application, you want to allow some grace to avoid immediate removal, as being too trigger happy and removing right may cause problems for example when the user did uninstall followed by fresh install.

To sum up, the question is: which of the 3 interpretations is the crrect one?

Thanks,
Shlomi

**Comment by jchambers** [comment]
Bleh. Good catch. I think this is a case of the upstream definition having changed a few times, and our docs only partially changing to match. When in doubt, the upstream docs are correct. We'll update the docs for the next release.

**Comment by hzshlomi** [comment]
Yeah, Obviously.. Just adding my 2 cents.. so, FYI, if i am not mistaken, you don't really get the 410 + Unregistered but only after a week or so has passed, and the invalidation timestamp is also always within a week time span from the time of the request.
Do you have a different experience with the 410? Did you ever notice a difference for example between voip push tokens and non-voip push tokens? silent push ("aps":{"content-available":1} is said to be throttled..)?

**Comment by batmstd** [comment]
After a week, the date will return not null in response? in fact, deleted the token, but the date returns null.

**Comment by jchambers** [comment]
> if i am not mistaken, you don't really get the 410 + Unregistered but only after a week or so has passed, and the invalidation timestamp is also always within a week time span from the time of the request.

That certainly seems plausible. The behavior has certainly changed over time. I suspect some fuzziness is a deliberate attempt to protect end-user privacy.

> Do you have a different experience with the 410? Did you ever notice a difference for example between voip push tokens and non-voip push tokens? silent push ("aps":{"content-available":1} is said to be throttled..)?

Sorry; I don't have any information to offer here.

**Comment by fzyzcjy** [comment]
@jchambers Sorry I am completely confused. I do not see anything related in the [apple doc](https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/sending_notification_requests_to_apns#2947616) provided from [the code](https://github.com/jchambers/pushy/blob/bf466c5236f3a7e29474df1b460858a2f7042b43/pushy/src/main/java/com/eatthepath/pushy/apns/PushNotificationResponse.java#L90). Thus, does it mean this invalidation timestamp *will never happen* now? Or do I understand something wrongly?

Thanks very much!

**Comment by fzyzcjy** [comment]
@jchambers I am a little bit confused. What does the invalidation timestamp actually tells? If the token is invalid, then it is invalid. If it is valid, then it is valid. But why it happens that "it is invalid, unless you get that token more recent than tokenInvalidationTimestamp"? Thanks!
